### PR TITLE
MODINVOICE-81 Implement API for invoice attachments (links/documents)

### PR DESCRIPTION
**Pull request by grabsefx**
## Purpose
https://issues.folio.org/browse/MODINVOICE-81

## Approach
* Implemented API
* Permissions added
* Implemented unit tests

#### TODOS and Open Questions
- [x] Merge PR https://github.com/folio-org/mod-invoice-storage/pull/33

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: ramls/invoice.raml [hunk 1]
```diff
@@ -64,6 +67,33 @@ resourceTypes:
         is: [validate]
       delete:
         description: Delete an invoice with given {id}
+
+      /documents:
+        displayName: Document
+        description: Manage documents assoctiated with invoice
```
Comment (piotr-kalashuk) [inline ramls/invoice.raml]: `assoctiated` -> `associated`

File: src/main/java/org/folio/rest/impl/DocumentHelper.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import me.escoffier.vertx.completablefuture.VertxCompletableFuture;
+
+class DocumentHelper extends AbstractHelper {
+
+  DocumentHelper(Map<String, String> okapiHeaders, Context ctx, String lang) {
+    super(getHttpClient(okapiHeaders), okapiHeaders, ctx, lang);
+  }
+
+  CompletableFuture<InvoiceDocument> createDocument(String invoiceId, InvoiceDocument document) {
+    JsonObject jsonDocument = JsonObject.mapFrom(document);
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/DocumentHelper.java]: According to jira desired approach is 
> If possible, reading the entire file's contents into memory should be avoided. A streaming approach would be ideal if feasible. 

As far as I can see RMB provides streaming capabilities only for [file uploads](https://github.com/folio-org/raml-module-builder/tree/master#file-uploads-option-2) where `application/octet-stream` is expected. So looks like there is no easy enough way to implement streaming approach...

File: src/main/java/org/folio/rest/impl/InvoicesImpl.java [hunk 3]
```diff
@@ -167,7 +170,57 @@ public void getInvoiceInvoiceNumber(String lang, Map<String, String> okapiHeader
...
+        })
+        .exceptionally(t -> handleErrorResponse(asyncResultHandler, documentHelper, t));
+    }
+  }
+
+  @Validate
+  @Override
+  public void getInvoiceInvoicesDocumentsById(String id, int offset, int limit, String query, String lang, Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    DocumentHelper documentHelper = new DocumentHelper(okapiHeaders, vertxContext, lang);
+    documentHelper.getDocumentsByInvoiceId(id)
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/InvoicesImpl.java]: The `query` should be taken into account (as well as `offset` and `limit`).

File: src/test/java/org/folio/rest/impl/MockServer.java [hunk 4]
```diff
@@ -5,47 +5,23 @@
...
-import static org.folio.invoices.utils.ResourcePathResolver.INVOICES;
-import static org.folio.invoices.utils.ResourcePathResolver.INVOICE_LINES;
-import static org.folio.invoices.utils.ResourcePathResolver.INVOICE_LINE_NUMBER;
-import static org.folio.invoices.utils.ResourcePathResolver.ORDER_LINES;
-import static org.folio.invoices.utils.ResourcePathResolver.VOUCHERS;
-import static org.folio.invoices.utils.ResourcePathResolver.VOUCHER_LINES;
-import static org.folio.invoices.utils.ResourcePathResolver.VOUCHER_NUMBER;
-import static org.folio.invoices.utils.ResourcePathResolver.VOUCHER_NUMBER_START;
-import static org.folio.invoices.utils.ResourcePathResolver.resourcesPath;
+import static org.folio.invoices.utils.ResourcePathResolver.*;
```
Comment (piotr-kalashuk) [inline src/test/java/org/folio/rest/impl/MockServer.java]: According to agreed [code style settings](https://wiki.folio.org/display/FOLIJET/Java+code+formatter+for+IDEs#JavacodeformatterforIDEs-OrganizeImports), the wildcard s going to be used for:
> Number of imports needed for .\*: 99
> Number of static imports needed for .\*: 99

File: ramls/invoice.raml [hunk 5]
```diff
@@ -64,6 +67,33 @@ resourceTypes:
         is: [validate]
       delete:
         description: Delete an invoice with given {id}
+
+      /documents:
```
Comment (piotr-kalashuk) [inline ramls/invoice.raml]: Please also use `validate` trait wherever useful...

File: src/main/java/org/folio/rest/impl/DocumentHelper.java (hunk 2 above)
Comment (craigmcnally) [reply src/main/java/org/folio/rest/impl/DocumentHelper.java]: This is probably fine for the initial implementation.  If we find it's an issue we might consider changing the API to accept an application/octet-stream, e.g. separate endpoints for uploading metadata and file contents.

File: ramls/invoice.raml (hunk 1 above)
Comment (grabsefx) [reply ramls/invoice.raml]: fixed typo

File: src/main/java/org/folio/rest/impl/InvoicesImpl.java (hunk 3 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/impl/InvoicesImpl.java]: implementation fixed

File: src/test/java/org/folio/rest/impl/MockServer.java (hunk 4 above)
Comment (grabsefx) [reply src/test/java/org/folio/rest/impl/MockServer.java]: formatted according to eclipse code style

File: ramls/invoice.raml (hunk 5 above)
Comment (grabsefx) [reply ramls/invoice.raml]: trait added

Comment (piotr-kalashuk) [reply ramls/invoice.raml]: I think better to add `validate` trait only to `POST`

File: src/main/java/org/folio/rest/impl/DocumentHelper.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    String endpoint = String.format(resourcesPath(ResourcePathResolver.INVOICE_DOCUMENTS), invoiceId);
+    return createRecordInStorage(jsonDocument, endpoint).thenApply(id -> {
+      document.getDocumentMetadata().setId(id);
+      return document;
+    });
+  }
+
+  CompletableFuture<DocumentCollection> getDocumentsByInvoiceId(String invoiceId, int limit, int offset, String query) {
+    CompletableFuture<DocumentCollection> future = new VertxCompletableFuture<>(ctx);
+    String queryParam = buildQuery(query, logger);
```
Comment (piotr-kalashuk) [inline src/main/java/org/folio/rest/impl/DocumentHelper.java]: Please use `HelperUtils.getEndpointWithQuery` instead. `buildQuery` method is duplicate of that one and I am going to remove it in https://github.com/folio-org/mod-invoice/pull/55

**Pull request merged**